app/models/user.rb

```diff
@@ -418,6 +421,54 @@ def setup(opts)
     self
   end
 
+  def is_pam_account?
+    return false unless Devise.pam_authentication
+    name = pam_get_name
+    return false unless name
+    Rpam2.account(Devise.pam_default_service, name).present?
+  end
+
+  def no_pam_conflict
+    errors.add(:username, 'username conflicts with pam account') if pam_conflict?
+  end
+
+  def pam_conflict?
+    encrypted_password.present? && is_pam_account?
+  end
+
+  def pam_conflict(_attributes)
+    # block pam logins on users with pw
+    nil
+  end
+
+  def pam_setup(attributes)
+    args = {}
+    args['username'] = self[::Devise.usernamefield]
+    args['email'] = Rpam2.getenv(get_service, self[::Devise.usernamefield], attributes[:password], "email", false)
+    args['email'] = "#{self[::Devise.usernamefield]}@#{find_pam_suffix}" if args['email'].nil?
+    opts = ActionController::Parameters.new(args)
+    setup(opts.permit(:username, :email))
+  end
+
+  def find_pam_suffix
+    "#{AppConfig.bare_pod_uri}"
+  end
+
+  def password_required?
+    return !is_pam_account? if Devise.pam_authentication
+    super
+  end
+
+  def send_reset_password_instructions
```

Comment:
Method `User#send_reset_password_instructions` is defined at both app/models/user.rb:173 and app/models/user.rb:462.